android/envoy/src/main/kotlin/org/greatfire/envoy/NetworkIntentService.kt

```diff
@@ -759,6 +769,74 @@ class NetworkIntentService : IntentService("NetworkIntentService") {
         }
     }
 
+    private fun handleMeekSubmit(
+        url: String,
+        captive_portal_url: String,
+        hysteriaCert: String?
+    ) {
+
+        val uri = URI(url)
+        var meekUrl = ""
+        var meekFront = ""
+        var meekTunnel = ""
+        val rawQuery = uri.rawQuery
+        val queries = rawQuery.split("&")
+        for (i in 0 until queries.size) {
+            val queryParts = queries[i].split("=")
+            if (queryParts[0].equals("url")) {
+                meekUrl = URLDecoder.decode(queryParts[1], "UTF-8")
+            } else if (queryParts[0].equals("front")) {
+                meekFront = URLDecoder.decode(queryParts[1], "UTF-8")
+                // if needed, generate randomized host name prefix
+                if (meekFront.startsWith('.')) {
+                    meekFront = randomString().plus(meekFront)
+                    Log.d(TAG, "front updated with random prefix: " + meekFront)
+                } else {
+                    Log.d(TAG, "front included as-is: " + meekFront)
+                }
+            } else if (queryParts[0].equals("tunnel")) {
+                // this is purposfully not decocded to add to an Envoy URL
+                meekTunnel = queryParts[1]
+            }
+        }
+        // start meek service
+        if (meekUrl.isNullOrEmpty() || meekFront.isNullOrEmpty() || meekTunnel.isNullOrEmpty()) {
+            Log.e(TAG, "some arguments required for meek service are missing")
+        } else {
+
+            // set the login params
+            val meekUserString = "url=" + meekUrl + ";front=" + meekFront
+            val nullCharString = "" + Char.MIN_VALUE
+
+            // additional hardcoded meek parameters
+            val logLevel = "DEBUG"
+            val enableLogging = true
+            val unsafeLogging = true
+
+            val meekPort = IEnvoyProxy.startMeek(meekUserString, nullCharString, logLevel, enableLogging, unsafeLogging)
+            val urlString = MEEK_URL_BASE_1 + meekTunnel + MEEK_URL_BASE_2 + meekPort
+
+            Log.d(TAG, "meek service started at " + urlString)
+
+            meekUrls.add(urlString)
+
+            // method returns port immediately but service is not ready immediately
+            Log.d(TAG, "submit url after a short delay for starting meek")
+            ioScope.launch() {
+                Log.d(TAG, "start meek delay")
+                delay(SHORT_DELAY)
```

Comment:
what if Meek isn't ready yet?

Reply:
the connection to the local proxy will fail, but during testing it has been ready within 1 second.

Reply:
tried lowering SHORT_DELAY to 500ms but then snowflake failed.  would rather not have a different delay for every service and i'm not sure 500ms is a big gain.  can revisit this approach later but the main difficulty is that iEnvoyProxy is calling a binary and i'm not aware of any way to get the feedback from that we'd need for more precise timing.

Reply:
500 ms is a pretty big chunk of time in terms of ui response, but let's not change this too much now, follow on funding we should definintely focus energy on how to implement an exponential backoff